code/game/machinery/vending.dm

```diff
@@ -305,9 +305,14 @@
 			attack_hand(user)
 		return
 	else if(istype(W, /obj/item/weapon/coin) && premium.len > 0)
-		user.drop_item(src)
-		coin = W
-		user << "<span class='notice'>You insert [W] into [src].</span>"
+		if (isnull(coin))
+			user.drop_item()

```

Comment:
Then I suggest you use drop_item(coinbox), or create some functionality in drop_item to drop to null space.
Better yet, have drop_item return the item dropped and do W = user.drop_item()
Assuming the item dropped by the user is always the same as the item used in Click code isn't always safe.